lib/mount.js

```diff
@@ -1,57 +1,42 @@
-const etag       = require('etag')
-const { Stream } = require('stream')
-
-const cleanRequest = require('./clean-request')
-const error        = require('./error')
-const getBody      = require('./get-body')
-const parseCookies = require('./parse-cookies')
-const parseUrl     = require('./parse-url')
-const { log, rethrow } = require('./util')
-
-const { byteLength, isBuffer } = Buffer
-
-const mount = (app, { errLogger, logger }={}) =>
-  (req, res) =>
+const { compose, identity, is, when, tap } = require('ramda')
+
+const cleanRequest  = require('./cleanRequest')
+const defaultLogger = require('./logger')
+const error         = require('./error')
+const getBody       = require('./getBody')
+const parseCookies  = require('./parseCookies')
+const parseUrl      = require('./parseUrl')
+const wrap          = require('./wrap')
+const write         = require('./write')
+
+const mount = (opts={}) => {
+  const {
+    app        = identity,
+    cry        = defaultLogger,
+    logger     = defaultLogger,
+    middleware = []
+  } = opts
+
+  const handle = middleware.length ? wrap(middleware, app) : app
+
+  const requestListener = (req, res) =>
     Promise.resolve(req)
       .then(getBody)
       .then(parseUrl)
       .then(parseCookies)
       .then(cleanRequest)
-      .then(app)
-      .catch(rethrow(errLogger))
-      .catch(error)
+      .then(handle)
+      .catch(when(
+        is(Error),
+        compose(error, tap(sob(req)))
```

Comment:
The reason for the `when(is(Error), ...)` is that sometimes in `paperplane` we do intentionally `reject` with an actual Response, not an Error.  The most common example of this is to `Promise.reject` early with a `redirect` response, skipping the rest of the app flow and responding early with a redirect.  This technique is even [mentioned in the docs](https://github.com/articulate/paperplane/blob/master/docs/API.md#redirect), along with an example.

I'd be interested to know what third-party lib you had trouble with, though.

Reply:
Oh, it's a _feature_. :confused: I never would have guessed. "Reject" === "Error" in my mind.

The lib where i encountered this was the [chargebee sdk](https://apidocs.chargebee.com/docs/api?lang=node). It uses node-style callbacks which i then promise-ified. But it's first `err` argument is not an `Error` type but a plain object. So when chargebee started responding with errors, paperplane responded with success with a malformed response object, & everything went :boom:

Reply:
Yuck.  So did you ended doing a `bimap` (or some `Promise` equivalent)?

Reply:
I ended up reinventing the wheel a bit.

```js
new Promise((resolve, reject) => {
  chargebeeFn((err, result) => {
    if (err != null) {
      reject(Object.assign(new Error(), err))
    }
    resolve(result)
  }
}
```

Reply:
Bummer.  `funky.reject` will lift a value into an `Error` if it isn't one already, but I'm not sure it would work for objects:
https://github.com/articulate/funky/blob/master/src/reject.js#L7